#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/competitions/desafio-de-prediccion-de-precios-de-portatiles

## Librerias

In [1]:
import pandas as pd


## Data

In [2]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 

train_data = pd.read_csv("desafio-de-prediccion-de-precios-de-portatiles/train.csv")

## Data exploration

In [3]:
train_data.head(20)

id  laptop_ID  Company                                       Product  \
0    268       1002     Dell                                 Inspiron 5567   
1    347        867     Asus                    X541NA (N4200/4GB/1TB/W10)   
2    353        966  Toshiba                             Portege Z30-C-1CW   
3    578        767     Dell                                  Alienware 17   
4     42       1241     Dell                                Latitude E7270   
5    867        376     Asus                                  ZenBook Flip   
6   1136         60     Asus  X541UA-DM1897 (i3-6006U/4GB/256GB/FHD/Linux)   
7    609       1157      MSI                                    GP62M 7RDX   
8    474        110     Asus    UX430UQ-GV209R (i7-7500U/8GB/256GB/GeForce   
9    838       1263     Asus                            Rog GL553VE-FY052T   
10   851       1243     Asus            X540SA-RBPDN09 (N3710/4GB/1TB/W10)   
11   904         79     Dell                                 Inspiron 5570   
12   355        957  Toshiba                               Tecra A50-C-1ZV   
13    44        579       HP          15-cd005nv (A9-9420/6GB/256GB/Radeon   
14    36        834  Toshiba                                 Satellite Pro   
15   971         82     Dell                                 Latitude 5590   
16   938        159   Lenovo                                Yoga 520-14IKB   
17   718        332     Acer                                 Aspire E5-575   
18   518        617   Lenovo                                  Thinkpad P51   
19  1126        172     Acer                               Aspire A517-51G   

              TypeName  Inches                           ScreenResolution  \
0             Notebook    15.6                                   1366x768   
1             Notebook    15.6                                   1366x768   
2             Notebook    13.3                          Full HD 1920x1080   
3               Gaming    15.6            IPS Panel 4K Ultra HD 3840x2160   
4            Ultrabook    12.5            Full HD / Touchscreen 1920x1080   
5   2 in 1 Convertible    15.6            Full HD / Touchscreen 1920x1080   
6             Notebook    15.6                          Full HD 1920x1080   
7               Gaming    15.6                          Full HD 1920x1080   
8            Ultrabook    14.0                IPS Panel Full HD 1920x1080   
9               Gaming    15.6                IPS Panel Full HD 1920x1080   
10            Notebook    15.6                                   1366x768   
11            Notebook    15.6                          Full HD 1920x1080   
12            Notebook    15.6                                   1366x768   
13            Notebook    15.6                IPS Panel Full HD 1920x1080   
14            Notebook    14.0                                   1366x768   
15           Ultrabook    15.6                IPS Panel Full HD 1920x1080   
16  2 in 1 Convertible    14.0  IPS Panel Full HD / Touchscreen 1920x1080   
17            Notebook    15.6                                   1366x768   
18            Notebook    15.6            IPS Panel 4K Ultra HD 3840x2160   
19            Notebook    17.3                IPS Panel Full HD 1920x1080   

                                     Cpu   Ram                Memory  \
0             Intel Core i7 7500U 2.7GHz   8GB               1TB HDD   
1   Intel Pentium Quad Core N4200 1.1GHz   4GB               1TB HDD   
2             Intel Core i5 6200U 2.3GHz   8GB             256GB SSD   
3            Intel Core i7 7700HQ 2.8GHz  16GB  256GB SSD +  1TB HDD   
4             Intel Core i5 6300U 2.4GHz   8GB             256GB SSD   
5             Intel Core i7 7500U 2.7GHz  12GB  512GB SSD +  2TB HDD   
6               Intel Core i3 6006U 2GHz   4GB             256GB SSD   
7            Intel Core i7 7700HQ 2.8GHz   8GB  128GB SSD +  1TB HDD   
8             Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   
9            Intel Core i7 77

In [4]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                912 non-null    int64  
 1   laptop_ID         912 non-null    int64  
 2   Company           912 non-null    object 
 3   Product           912 non-null    object 
 4   TypeName          912 non-null    object 
 5   Inches            912 non-null    float64
 6   ScreenResolution  912 non-null    object 
 7   Cpu               912 non-null    object 
 8   Ram               912 non-null    object 
 9   Memory            912 non-null    object 
 10  Gpu               912 non-null    object 
 11  OpSys             912 non-null    object 
 12  Weight            912 non-null    object 
 13  Price_euros       912 non-null    float64
dtypes: float64(2), int64(2), object(10)
memory usage: 99.9+ KB


In [5]:
train_data.describe()

id    laptop_ID      Inches  Price_euros
count   912.000000   912.000000  912.000000   912.000000
mean    645.091009   666.192982   15.011404  1108.122873
std     380.313813   384.873846    1.411744   714.597741
min       0.000000     1.000000   10.100000   174.000000
25%     319.750000   330.500000   14.000000   589.000000
50%     636.500000   673.500000   15.600000   949.000000
75%     980.250000   998.500000   15.600000  1458.500000
max    1302.000000  1320.000000   18.400000  6099.000000

In [6]:
sample_data = pd.read_csv("desafio-de-prediccion-de-precios-de-portatiles/sample_submission.csv")
sample_data

id  Price_euros
0    1014        752.0
1     845        499.0
2    1151       1747.0
3    1265        245.0
4     573       1179.0
..    ...          ...
386  1162        258.0
387  1286       1085.0
388   336        298.0
389   885       1099.0
390  1246       1339.0

[391 rows x 2 columns]

## Data processing

In [7]:
train_data['Inches'] = train_data['Inches'].astype(float)
train_data['Weight'] = train_data['Weight'].str.replace('kg', '').astype(float)


In [8]:
train_data.drop(columns=['laptop_ID'], inplace=True)


In [9]:
train_data.head()

id  Company                     Product   TypeName  Inches  \
0  268     Dell               Inspiron 5567   Notebook    15.6   
1  347     Asus  X541NA (N4200/4GB/1TB/W10)   Notebook    15.6   
2  353  Toshiba           Portege Z30-C-1CW   Notebook    13.3   
3  578     Dell                Alienware 17     Gaming    15.6   
4   42     Dell              Latitude E7270  Ultrabook    12.5   

                  ScreenResolution                                   Cpu  \
0                         1366x768            Intel Core i7 7500U 2.7GHz   
1                         1366x768  Intel Pentium Quad Core N4200 1.1GHz   
2                Full HD 1920x1080            Intel Core i5 6200U 2.3GHz   
3  IPS Panel 4K Ultra HD 3840x2160           Intel Core i7 7700HQ 2.8GHz   
4  Full HD / Touchscreen 1920x1080            Intel Core i5 6300U 2.4GHz   

    Ram                Memory                      Gpu       OpSys  Weight  \
0   8GB               1TB HDD       AMD Radeon R7 M445  Windows 10    2.36   
1   4GB               1TB HDD    Intel HD Graphics 505  Windows 10    2.00   
2   8GB             256GB SSD    Intel HD Graphics 520   Windows 7    1.20   
3  16GB  256GB SSD +  1TB HDD  Nvidia GeForce GTX 1070  Windows 10    4.42   
4   8GB             256GB SSD    Intel HD Graphics 520   Windows 7    1.26   

   Price_euros  
0       749.00  
1       449.00  
2      1460.00  
3      2868.99  
4      1713.37

In [10]:
# La columna Ram contiene valores en formato de cadena con 'GB' al final. Vamos a remover 'GB' y convertir los valores a numéricos.
train_data['Ram'] = train_data['Ram'].str.replace('GB', '').astype(int)

In [11]:
train_data.head()

id  Company                     Product   TypeName  Inches  \
0  268     Dell               Inspiron 5567   Notebook    15.6   
1  347     Asus  X541NA (N4200/4GB/1TB/W10)   Notebook    15.6   
2  353  Toshiba           Portege Z30-C-1CW   Notebook    13.3   
3  578     Dell                Alienware 17     Gaming    15.6   
4   42     Dell              Latitude E7270  Ultrabook    12.5   

                  ScreenResolution                                   Cpu  Ram  \
0                         1366x768            Intel Core i7 7500U 2.7GHz    8   
1                         1366x768  Intel Pentium Quad Core N4200 1.1GHz    4   
2                Full HD 1920x1080            Intel Core i5 6200U 2.3GHz    8   
3  IPS Panel 4K Ultra HD 3840x2160           Intel Core i7 7700HQ 2.8GHz   16   
4  Full HD / Touchscreen 1920x1080            Intel Core i5 6300U 2.4GHz    8   

                 Memory                      Gpu       OpSys  Weight  \
0               1TB HDD       AMD Radeon R7 M445  Windows 10    2.36   
1               1TB HDD    Intel HD Graphics 505  Windows 10    2.00   
2             256GB SSD    Intel HD Graphics 520   Windows 7    1.20   
3  256GB SSD +  1TB HDD  Nvidia GeForce GTX 1070  Windows 10    4.42   
4             256GB SSD    Intel HD Graphics 520   Windows 7    1.26   

   Price_euros  
0       749.00  
1       449.00  
2      1460.00  
3      2868.99  
4      1713.37

In [12]:
# Crear funciones para extraer la resolución. La columna ScreenResolution tiene resoluciones en diferentes formatos, incluyendo menciones de paneles táctiles. Vamos a separarla en múltiples columnas.
def get_x_resolution(x):
    return int(x.split('x')[0].split()[-1])

def get_y_resolution(x):
    return int(x.split('x')[1].split()[0])

train_data['X_resolution'] = train_data['ScreenResolution'].apply(get_x_resolution)
train_data['Y_resolution'] = train_data['ScreenResolution'].apply(get_y_resolution)
train_data.drop(columns=['ScreenResolution'], inplace=True)


In [13]:
train_data.head()

id  Company                     Product   TypeName  Inches  \
0  268     Dell               Inspiron 5567   Notebook    15.6   
1  347     Asus  X541NA (N4200/4GB/1TB/W10)   Notebook    15.6   
2  353  Toshiba           Portege Z30-C-1CW   Notebook    13.3   
3  578     Dell                Alienware 17     Gaming    15.6   
4   42     Dell              Latitude E7270  Ultrabook    12.5   

                                    Cpu  Ram                Memory  \
0            Intel Core i7 7500U 2.7GHz    8               1TB HDD   
1  Intel Pentium Quad Core N4200 1.1GHz    4               1TB HDD   
2            Intel Core i5 6200U 2.3GHz    8             256GB SSD   
3           Intel Core i7 7700HQ 2.8GHz   16  256GB SSD +  1TB HDD   
4            Intel Core i5 6300U 2.4GHz    8             256GB SSD   

                       Gpu       OpSys  Weight  Price_euros  X_resolution  \
0       AMD Radeon R7 M445  Windows 10    2.36       749.00          1366   
1    Intel HD Graphics 505  Windows 10    2.00       449.00          1366   
2    Intel HD Graphics 520   Windows 7    1.20      1460.00          1920   
3  Nvidia GeForce GTX 1070  Windows 10    4.42      2868.99          3840   
4    Intel HD Graphics 520   Windows 7    1.26      1713.37          1920   

   Y_resolution  
0           768  
1           768  
2          1080  
3          2160  
4          1080

In [14]:
# Vamos a separar la información de la CPU en tres columnas: marca, serie y velocidad.
def extract_cpu_info(cpu):
    parts = cpu.split()
    brand = parts[0]
    series = " ".join(parts[1:-1])
    speed = parts[-1].replace('GHz', '')
    return brand, series, speed

cpu_info = train_data['Cpu'].apply(lambda x: pd.Series(extract_cpu_info(x)))
cpu_info.columns = ['Cpu_Brand', 'Cpu_Series', 'Cpu_Speed']
cpu_info['Cpu_Speed'] = cpu_info['Cpu_Speed'].astype(float)

train_data = pd.concat([train_data, cpu_info], axis=1)
train_data.drop(columns=['Cpu'], inplace=True)


In [15]:
train_data.head()

id  Company                     Product   TypeName  Inches  Ram  \
0  268     Dell               Inspiron 5567   Notebook    15.6    8   
1  347     Asus  X541NA (N4200/4GB/1TB/W10)   Notebook    15.6    4   
2  353  Toshiba           Portege Z30-C-1CW   Notebook    13.3    8   
3  578     Dell                Alienware 17     Gaming    15.6   16   
4   42     Dell              Latitude E7270  Ultrabook    12.5    8   

                 Memory                      Gpu       OpSys  Weight  \
0               1TB HDD       AMD Radeon R7 M445  Windows 10    2.36   
1               1TB HDD    Intel HD Graphics 505  Windows 10    2.00   
2             256GB SSD    Intel HD Graphics 520   Windows 7    1.20   
3  256GB SSD +  1TB HDD  Nvidia GeForce GTX 1070  Windows 10    4.42   
4             256GB SSD    Intel HD Graphics 520   Windows 7    1.26   

   Price_euros  X_resolution  Y_resolution Cpu_Brand               Cpu_Series  \
0       749.00          1366           768     Intel            Core i7 7500U   
1       449.00          1366           768     Intel  Pentium Quad Core N4200   
2      1460.00          1920          1080     Intel            Core i5 6200U   
3      2868.99          3840          2160     Intel           Core i7 7700HQ   
4      1713.37          1920          1080     Intel            Core i5 6300U   

   Cpu_Speed  
0        2.7  
1        1.1  
2        2.3  
3        2.8  
4        2.4

In [16]:
# La columna Memory contiene diferentes configuraciones de almacenamiento (HDD, SSD, combinaciones). Vamos a separar esta información en varias columnas.
import re

def parse_memory(memory):
    hdd = 0
    ssd = 0
    flash = 0
    hybrid = 0

    if 'TB' in memory:
        memory = memory.replace('TB', '000GB')
    
    for part in memory.split('+'):
        part = part.strip()
        if 'GB' in part:
            size = int(re.findall(r'\d+', part)[0])
            if 'SSD' in part:
                ssd += size
            elif 'HDD' in part:
                hdd += size
            elif 'Hybrid' in part:
                hybrid += size
            elif 'Flash Storage' in part:
                flash += size

    return hdd, ssd, flash, hybrid

memory_info = train_data['Memory'].apply(lambda x: pd.Series(parse_memory(x)))
memory_info.columns = ['HDD', 'SSD', 'Flash_Storage', 'Hybrid']

train_data = pd.concat([train_data, memory_info], axis=1)
train_data.drop(columns=['Memory'], inplace=True)


In [34]:
train_data.head()

id  Company                     Product   TypeName  Inches  Ram  \
0  268     Dell               Inspiron 5567   Notebook    15.6    8   
1  347     Asus  X541NA (N4200/4GB/1TB/W10)   Notebook    15.6    4   
2  353  Toshiba           Portege Z30-C-1CW   Notebook    13.3    8   
3  578     Dell                Alienware 17     Gaming    15.6   16   
4   42     Dell              Latitude E7270  Ultrabook    12.5    8   

                       Gpu       OpSys  Weight  Price_euros  X_resolution  \
0       AMD Radeon R7 M445  Windows 10    2.36       749.00          1366   
1    Intel HD Graphics 505  Windows 10    2.00       449.00          1366   
2    Intel HD Graphics 520   Windows 7    1.20      1460.00          1920   
3  Nvidia GeForce GTX 1070  Windows 10    4.42      2868.99          3840   
4    Intel HD Graphics 520   Windows 7    1.26      1713.37          1920   

   Y_resolution Cpu_Brand               Cpu_Series  Cpu_Speed   HDD  SSD  \
0           768     Intel            Core i7 7500U        2.7  1000    0   
1           768     Intel  Pentium Quad Core N4200        1.1  1000    0   
2          1080     Intel            Core i5 6200U        2.3     0  256   
3          2160     Intel           Core i7 7700HQ        2.8  1000  256   
4          1080     Intel            Core i5 6300U        2.4     0  256   

   Flash_Storage  Hybrid  
0              0       0  
1              0       0  
2              0       0  
3              0       0  
4              0       0

In [17]:
# Las columnas categóricas deben ser convertidas a variables dummy para ser utilizadas en el modelo
categorical_columns = ['Company', 'Product', 'TypeName', 'Gpu', 'OpSys', 'Cpu_Brand', 'Cpu_Series']

train_data = pd.get_dummies(train_data, columns=categorical_columns)


In [18]:
# Guardamos el conjunto de datos limpio en un nuevo archivo CSV para su uso
train_data.to_csv("desafio-de-prediccion-de-precios-de-portatiles/train_clean.csv", index=False)


-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [19]:
# Cargar el conjunto de datos limpio
train_clean = pd.read_csv("desafio-de-prediccion-de-precios-de-portatiles/train_clean.csv")

# Definir X e y
X = train_clean.drop(columns=['Price_euros'])
y = train_clean['Price_euros']

# Verificar las primeras filas de X e y
print(X.head())
print(y.head())


    id  Inches  Ram  Weight  X_resolution  Y_resolution  Cpu_Speed   HDD  SSD  \
0  268    15.6    8    2.36          1366           768        2.7  1000    0   
1  347    15.6    4    2.00          1366           768        1.1  1000    0   
2  353    13.3    8    1.20          1920          1080        2.3     0  256   
3  578    15.6   16    4.42          3840          2160        2.8  1000  256   
4   42    12.5    8    1.26          1920          1080        2.4     0  256   

   Flash_Storage  ...  Cpu_Series_Pentium Dual Core 4405Y  \
0              0  ...                               False   
1              0  ...                               False   
2              0  ...                               False   
3              0  ...                               False   
4              0  ...                               False   

   Cpu_Series_Pentium Dual Core N4200  Cpu_Series_Pentium Quad Core N3700  \
0                               False                               F

En X, tenemos:

Inches: Tamaño de la pantalla en pulgadas.
Ram: Memoria RAM en GB.
Weight: Peso del portátil en kg.
X_resolution: Resolución horizontal de la pantalla.
Y_resolution: Resolución vertical de la pantalla.
Cpu_Speed: Velocidad del procesador en GHz.
HDD: Capacidad del disco duro en GB.
SSD: Capacidad del SSD en GB.
Flash_Storage: Capacidad del almacenamiento flash en GB.
Hybrid: Capacidad del almacenamiento híbrido en GB.
Variables Dummy: Incluyen todas las columnas de variables dummy para Company, Product, TypeName, Gpu, OpSys, Cpu_Brand, y Cpu_Series.

### 2. Dividir X_train, X_test, y_train, y_test

In [20]:
from sklearn.model_selection import train_test_split

# Definir X e y
X = train_clean.drop(columns=['Price_euros'])
y = train_clean['Price_euros']

# Dividir en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Verificar las dimensiones de los conjuntos
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)


X_train shape: (729, 696)
X_test shape: (183, 696)
y_train shape: (729,)
y_test shape: (183,)


### 3. Crear y entrenar el modelo


In [21]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Crear y entrenar el modelo de regresión lineal
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred_linear = linear_model.predict(X_test)

# Evaluar el desempeño del modelo
mae_linear = mean_absolute_error(y_test, y_pred_linear)
mse_linear = mean_squared_error(y_test, y_pred_linear)
r2_linear = r2_score(y_test, y_pred_linear)

print("Regresión Lineal")
print("MAE:", mae_linear)
print("MSE:", mse_linear)
print("R2 Score:", r2_linear)


Regresión Lineal
MAE: 593.9159641626949
MSE: 7075154.025922916
R2 Score: -15.38123190831789


In [22]:
from sklearn.ensemble import RandomForestRegressor

# Crear y entrenar el modelo de Random Forest
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred_rf = rf_model.predict(X_test)

# Evaluar el desempeño del modelo
mae_rf = mean_absolute_error(y_test, y_pred_rf)
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print("Random Forest")
print("MAE:", mae_rf)
print("MSE:", mse_rf)
print("R2 Score:", r2_rf)


Random Forest
MAE: 189.75958961748634
MSE: 100363.18465344736
R2 Score: 0.767627305802577


In [23]:
from sklearn.model_selection import GridSearchCV

# Definir el rango de hiperparámetros para buscar
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Crear el modelo de Random Forest
rf_model = RandomForestRegressor(random_state=42)

# Configurar la búsqueda en cuadrícula
grid_search = GridSearchCV(estimator=rf_model, param_grid=param_grid, 
                           cv=3, n_jobs=-1, verbose=2, scoring='neg_mean_squared_error')

# Ejecutar la búsqueda en cuadrícula
grid_search.fit(X_train, y_train)

# Obtener los mejores hiperparámetros
best_params = grid_search.best_params_
print("Mejores hiperparámetros:", best_params)

# Evaluar el mejor modelo en el conjunto de prueba
best_rf_model = grid_search.best_estimator_
y_pred_best_rf = best_rf_model.predict(X_test)

# Evaluar el desempeño del mejor modelo
mae_best_rf = mean_absolute_error(y_test, y_pred_best_rf)
mse_best_rf = mean_squared_error(y_test, y_pred_best_rf)
r2_best_rf = r2_score(y_test, y_pred_best_rf)

print("Random Forest Optimizado")
print("MAE:", mae_best_rf)
print("MSE:", mse_best_rf)
print("R2 Score:", r2_best_rf)


Fitting 3 folds for each of 108 candidates, totalling 324 fits
Mejores hiperparámetros: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 300}
Random Forest Optimizado
MAE: 190.00081985428045
MSE: 98102.54973453376
R2 Score: 0.7728613946621384


Explicación del Código
Definir el rango de hiperparámetros: Establecemos un rango de valores posibles para n_estimators, max_depth, min_samples_split, y min_samples_leaf.
Crear el modelo de Random Forest: Creamos una instancia de RandomForestRegressor.
Configurar la búsqueda en cuadrícula: Utilizamos GridSearchCV para buscar a través del espacio de hiperparámetros. Configuramos cv=3 para realizar una validación cruzada de 3 pliegues y n_jobs=-1 para utilizar todos los núcleos disponibles.
Ejecutar la búsqueda en cuadrícula: Ajustamos el modelo a los datos de entrenamiento.
Obtener los mejores hiperparámetros: Extraemos los mejores hiperparámetros encontrados durante la búsqueda.
Evaluar el mejor modelo: Utilizamos el mejor modelo encontrado para hacer predicciones en el conjunto de prueba y evaluamos su desempeño.

Próximos Pasos
Para mejorar aún más el modelo, se pueden considerar las siguientes acciones:

Exploración Adicional de Hiperparámetros: Probar con un rango más amplio de hiperparámetros y otras técnicas de búsqueda, como RandomizedSearchCV.

Más Datos y Features: Si es posible, obtener más datos y explorar nuevas características (features) que puedan tener una alta correlación con el precio.

Feature Engineering: Crear nuevas características derivadas de las existentes que puedan proporcionar mejor información al modelo.

Modelos Alternativos: Probar otros modelos de machine learning como Gradient Boosting, XGBoost, o LightGBM, que a menudo pueden proporcionar mejores resultados en tareas de regresión.

Ensemble Methods: Combinar múltiples modelos para aprovechar las fortalezas de cada uno y mejorar la precisión general.

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

**¿Por qué puede dar error?** 

IMPORTANTE: APLICAR A ESTOS DATOS LO MISMO QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [26]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [27]:
predictions = model.predict(X_pred)
predictions

NameError: name 'model' is not defined

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [ ]:
sample = pd.read_csv("Dataset/sample_submission.csv") # Esta es mi ruta del archivo, la vuestra puede ser otra

In [ ]:
sample.head()

In [ ]:
sample.shape

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [ ]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("")


In [ ]:
chequeator(submission) # submission es el nombre que le habríamos puesto a nuestro .csv con los valores que me salieron en la predicción